wave_kinematics/include/wave/kinematics/numerical_test_functors.hpp

```diff
@@ -0,0 +1,78 @@
+#ifndef __WAVE_NUMERICAL_TEST_FUNCTORS_HPP__
+#define __WAVE_NUMERICAL_TEST_FUNCTORS_HPP__
+
+#include <iostream>
+#include <functional>
+
+#include "wave/kinematics/rotation.hpp"
+#include "wave/utils/utils.hpp"
+#include <kindr/Core>
+
+namespace wave {
+
+// The Functors are used for computing finite difference Jacobians
+// of rotation expressions.
+
+class RotateAndJacobianJpointFunctor {
+public:
+  wave::Rotation R;
+  wave::Vec3 P;
+  RotateAndJacobianJpointFunctor(wave::Rotation input_rotation) {
+    this->R = input_rotation;
+  }
+
+  int operator()(wave::Vec3 input_point, wave::Vec3 &output_point) {
+    output_point = R.rotate(input_point);
+  }
+};
+
+
+class RotateAndJacobianJparamFunctor {
+public:
+  wave::Rotation R;
+  wave::Vec3 P;
+  RotateAndJacobianJparamFunctor(wave::Rotation input_rotation,
+                                 wave::Vec3 input_point) {
+    this->R = input_rotation;
+    this->P = input_point;
+  }
+
+  int operator()(wave::Vec3 input_point, wave::Vec3 &output_point) {
+    wave::Rotation Rp = this->R;
+    Rp.manifoldPlus(input_point);
+    output_point = Rp.rotate(P);
+  }
+};
+
+
+// Implement the numerical differentiation.  Based on the Eigen::NumericalDiff
+// Module.
```

Comment:
Yes. Eigen's MPL license is copyleft at file level. It looks like any files containing modified MPL-licensed code must be distributed under MPL. See [MPL 2.0 FAQ](https://www.mozilla.org/en-US/MPL/2.0/FAQ/)

A few possibilities:

* Put Eigen-derived code in a separate file and adhere to license
* Wipe your memory and implement the algorithm separately
* Use Eigen's `NumericalDiff.h` and define an `operator-` for your value type. It will be used.